tools/src/merger/OpenApiMerger.ts

```diff
@@ -59,21 +74,83 @@ export default class OpenApiMerger {
+      this.#redirect_refs_in_namespace(spec)
+      this._spec.paths = { ...this._spec.paths, ...spec.paths }
+      this._spec.components.parameters = { ...this._spec.components.parameters, ...spec.components.parameters }
+      this._spec.components.responses = { ...this._spec.components.responses, ...spec.components.responses }
+      this._spec.components.requestBodies = { ...this._spec.components.requestBodies, ...spec.components.requestBodies }
+    })
+
+    this.#remove_refs_per_semver()
+  }
+
+  // Remove any refs that are x-version-added/removed incompatible with the target server version.
+  #remove_refs_per_semver() : void {
+    this.#remove_per_semver(this._spec.paths)
+
+    // parameters
+    const removed_params = this.#remove_per_semver(this._spec.components.parameters)
+    const removed_parameter_refs = _.map(removed_params, (ref) => `#/components/parameters/${ref}`)
+    Object.entries(this._spec.paths as Document).forEach(([_path, path_item]) => {
+      Object.entries(path_item as Document).forEach(([_method, method_item]) => {
+        method_item.parameters = _.filter(method_item.parameters, (param) => !_.includes(removed_parameter_refs, param.$ref))
+      })
+    })
+
+    // responses
+    const removed_responses = this.#remove_per_semver(this._spec.components.responses)
+    const removed_response_refs = _.map(removed_responses, (ref) => `#/components/responses/${ref}`)
+    Object.entries(this._spec.paths as Document).forEach(([_path, path_item]) => {
+      Object.entries(path_item as Document).forEach(([_method, method_item]) => {
+        method_item.responses = _.omitBy(method_item.responses, (param) => _.includes(removed_response_refs, param.$ref))
+      })
     })
+
+    this._spec.paths = _.omitBy(this._spec.paths, isEmpty)
+  }
+
+  #exclude_per_semver(obj: any): boolean {
+    if (this.target_version === undefined) return false
+
+    const x_version_added = semver.coerce(obj['x-version-added'] as string)
+    const x_version_removed = semver.coerce(obj['x-version-removed'] as string)
+
+    if (x_version_added && !semver.satisfies(this.target_version, `>=${x_version_added?.toString()}`)) {
+      return true
+    } else if (x_version_removed && !semver.satisfies(this.target_version, `<${x_version_removed?.toString()}`)) {
+      return true
+    }
+
+    return false
+  }
+
+  // Remove any elements that are x-version-added/removed incompatible with the target server version.
+  #remove_per_semver(obj: any): string[] {
+    let removed: string[] = []
+    if (this.target_version === undefined) return removed
+
+    for (const key in obj) {
+      if (_.isObject(obj[key])) {
+        if (this.#exclude_per_semver(obj[key])) {
+          removed.push(key)
+          // eslint-disable-next-line @typescript-eslint/no-dynamic-delete
+          delete obj[key]
+        } else {
+          removed = _.concat(removed, this.#remove_per_semver(obj[key]))
+        }
+      }
+    }
+
+    return removed
```

Comment:
Isolation-of-concerns wise, I think this logic should be of its own class (for better readability and testing). The `OpenApiMerger`'s class' only duty is turning all the YAML files in the `spec` folder into a single OpenAPI yaml file without any modification. Then we have something like `ExtractSpecByVersion` class that uses the output of this  class to remove the irrelevant bits.

Reply:
Extracted into a helper.

Reply:
I meant this [entire code block](https://github.com/opensearch-project/opensearch-api-specification/pull/428/files#diff-707690e395baf72e22f4878dc5f5d802fc2a0a558ca4d56f9b49e67a8e3861fcR80-R126) should be its own class (say `ExtractSpecByVersion`) that consumes the output of the current `OpenApiMerger`. It makes the logic easier to follow and test as `OpenApiMerger` would be doing 2 jobs that intertwine with each other (merge all files in `./spec` and remove by `x-version...`).

Also, maybe I've missed it, but does this also remove items in `#components/schemas` that are no longer referenced after the removal of operation components?

Reply:
I see. I think I disagree with that. For me, merging a spec targeting a different version is just a feature of a merger, not something that takes an existing merged spec to create another merged spec. LMK if you feel strongly about this and I'll extract it.

Reply:
It's just a matter of opinion. It's easier to remove these components and the affected components (i.e. the unused items in `#components/schemas`) from a single-file spec than removing them from the individual spec file during merging and then figure out which other files are affected.

Reply:
I ended up extracting this into a class as you suggested, you're right, it's cleaner.